pkg/clients/ec2/subnet.go

```diff
@@ -33,3 +48,73 @@ func IsSubnetNotFoundErr(err error) bool {
 
 	return false
 }
+
+// GenerateSubnetObservation is used to produce v1beta1.SubnetExternalStatus from
+// ec2.Subnet
+func GenerateSubnetObservation(subnet ec2.Subnet) v1beta1.SubnetObservation {
+	o := v1beta1.SubnetObservation{
+		AvailableIPAddressCount: aws.Int64Value(subnet.AvailableIpAddressCount),
+		DefaultForAz:            aws.BoolValue(subnet.DefaultForAz),
+		SubnetID:                aws.StringValue(subnet.SubnetId),
+		SubnetState:             string(subnet.State),
+	}
+
+	v, err := subnet.State.MarshalValue()
+	if err != nil {
+		o.SubnetState = v
+	}
+
+	return o
+}
+
+// LateInitializeSubnet fills the empty fields in *v1beta1.SubnetParameters with
+// the values seen in ec2.Subnet.
+func LateInitializeSubnet(in *v1beta1.SubnetParameters, s *ec2.Subnet) { // nolint:gocyclo
+	if s == nil {
+		return
+	}
+
+	in.AssignIpv6AddressOnCreation = awsclients.LateInitializeBoolPtr(in.AssignIpv6AddressOnCreation, s.AssignIpv6AddressOnCreation)
+	in.AvailabilityZone = awsclients.LateInitializeStringPtr(in.AvailabilityZone, s.AvailabilityZone)
+	in.AvailabilityZoneID = awsclients.LateInitializeStringPtr(in.AvailabilityZoneID, s.AvailabilityZoneId)
+	in.CIDRBlock = awsclients.LateInitializeString(in.CIDRBlock, s.CidrBlock)
+	in.MapPublicIPOnLaunch = awsclients.LateInitializeBoolPtr(in.MapPublicIPOnLaunch, s.MapPublicIpOnLaunch)
+	in.VPCID = awsclients.LateInitializeStringPtr(in.VPCID, s.VpcId)
+
+	if s.Ipv6CidrBlockAssociationSet != nil {
+		in.Ipv6CIDRBlock = awsclients.LateInitializeStringPtr(in.Ipv6CIDRBlock, s.Ipv6CidrBlockAssociationSet[0].Ipv6CidrBlock)
+	}
+
+	if len(in.Tags) == 0 && len(s.Tags) != 0 {
+		in.Tags = v1beta1.BuildFromEC2Tags(s.Tags)
+	}
+}
+
+// CreateSubnetPatch creates a *v1beta1.SubnetParameters that has only the changed
+// values between the target *v1beta1.SubnetParameters and the current
+// *ec2.Subnet
+func CreateSubnetPatch(in ec2.Subnet, target v1beta1.SubnetParameters) (*v1beta1.SubnetParameters, error) {
+	currentParams := &v1beta1.SubnetParameters{}
+
+	v1beta1.SortTags(target.Tags, in.Tags)
+	LateInitializeSubnet(currentParams, &in)
+
+	jsonPatch, err := awsclients.CreateJSONPatch(currentParams, target)
+	if err != nil {
+		return nil, err
+	}
+	patch := &v1beta1.SubnetParameters{}
+	if err := json.Unmarshal(jsonPatch, patch); err != nil {
+		return nil, err
+	}
+	return patch, nil
+}
+
+// IsSubnetUpToDate checks whether there is a change in any of the modifiable fields.
+func IsSubnetUpToDate(p v1beta1.SubnetParameters, s ec2.Subnet) (bool, error) {
+	patch, err := CreateSubnetPatch(s, p)
```

Comment:
As far as I can see, there are only 2 fields that can be changed. So, I'd just do a `switch` statement for `IsUpToDate` check and get rid of `func CreateSubnetPatch`

Reply:
Same for calling `CreateSubnetPatch` in `Update`. It's barely a few fields, so, handling them manually makes it easier to grasp the code.